### ISSUE TITLE
"write() argument must be str, not bytes" or "[platform/windows/godot_res.windows.tools.64.o] Error 1" on windows via scons

### ISSUE THREAD

**Issue by MichaelBelousov** [godotengine/godot#38164]
**Godot version:**
3.2 branch, hash:  ed27b7e6
master 28f0b15c

**OS/device including version:**
Windows 10 Home 10.0.18362

**Issue description:**
SCons build fails with both msys/scons or pip-installed scons, on a fresh installation of msys2 using the mingw64 subsystem. Will post more env info as needed:

Python Version: (default msys2 as of today) 3.8.2
Msys explicitly installed packages:
```Bash
$ pacman -Qe
bash 4.4.023-2
bash-completion 2.10-1
bsdcpio 3.4.2-2
bsdtar 3.4.2-2
bzip2 1.0.8-2
coreutils 8.32-1
curl 7.69.1-1
dash 0.5.10.2-1
dtc 1.6.0-2
file 5.38-3
filesystem 2020.02-2
findutils 4.7.0-1
flex 2.6.4-1
gawk 5.1.0-1
gcc-libs 9.3.0-1
getent 2.18.90-2
git 2.26.2-1
grep 3.0-2
gzip 1.10-1
inetutils 1.9.4-2
info 6.7-1
less 551-1
libargp 20110921-2
lndir 1.0.3-1
make 4.3-1
man-db 2.9.1-1
mingw-w64-i686-gcc 9.3.0-2
mingw-w64-i686-python-pip 20.0.2-1
mingw-w64-i686-yasm 1.3.0-4
mingw-w64-x86_64-gcc 9.3.0-2
mingw-w64-x86_64-gcc-ada 9.3.0-2
mingw-w64-x86_64-gcc-fortran 9.3.0-2
mingw-w64-x86_64-gcc-libgfortran 9.3.0-2
mingw-w64-x86_64-gcc-objc 9.3.0-2
mingw-w64-x86_64-gdb 9.1-3
mingw-w64-x86_64-git-lfs 2.9.2-1
mingw-w64-x86_64-libmangle-git 8.0.0.5573.a4de8df4-1
mingw-w64-x86_64-make 4.3-1
mingw-w64-x86_64-pkg-config 0.29.2-1
mingw-w64-x86_64-python-pip 20.0.2-1
mingw-w64-x86_64-tools-git 8.0.0.5573.a4de8df4-1
mingw-w64-x86_64-winstorecompat-git 8.0.0.5573.a4de8df4-1
mingw-w64-x86_64-yasm 1.3.0-4
mintty 1~3.1.4-1
msys2-keyring r9.397a52e-1
msys2-launcher-git 0.3.32.56c2ba7-2
msys2-runtime 3.0.7-6
ncurses 6.2-1
pacman 5.2.1-6
pacman-mirrors 20200329-1
pactoys-git r2.07ca37f-1
pax-git 20161104.2-1
pkgfile 21-1
rebase 4.4.4-1
sed 4.8-1
tftp-hpa 5.2-3
time 1.9-1
tmux 3.0.a-1
ttyrec 1.0.8-2
tzcode 2019.a-1
util-linux 2.35.1-1
vim 8.2.0592-1
which 2.21-2
```

**Steps to reproduce:**
- install msys2
- clone godotengine/godot and checkout the 3.2 branch (also master)
- install packages from [the windows setup guide](https://docs.godotengine.org/en/stable/development/compiling/compiling_for_windows.html).
(notably:`mingw-w64-x86_64-python3-pip mingw-w64-x86_64-gcc mingw-w64-x86_64-yasm mingw-w64-i686-python3-pip mingw-w64-i686-gcc mingw-w64-i686-yasm make` 
- install scons via pip:
- `pip install scons` or `python -m pip install scons`
- build with scons with `scons platform=windows`
result:
```Bash
$ scons p=windows
scons: Reading SConscript files ...
TypeError: write() argument must be str, not bytes:
  File "D:/gamedev/godot/SConstruct", line 3:
    EnsureSConsVersion(0, 98, 1)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Script/SConscript.py", line 658:
    env = self.factory()
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Script/SConscript.py", line 638:
    default_env = SCons.Defaults.DefaultEnvironment()
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Defaults.py", line 88:
    _default_env = SCons.Environment.Environment(*args, **kw)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Environment.py", line 990:
    apply_tools(self, tools, toolpath)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Environment.py", line 102:
    env.Tool(tool)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Environment.py", line 1810:
    tool(self)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Tool/__init__.py", line 303:
    self.generate(env, *args, **kw)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Tool/default.py", line 41:
    SCons.Tool.Tool(t)(env)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Tool/__init__.py", line 303:
    self.generate(env, *args, **kw)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Tool/mslink.py", line 314:
    msvc_setup_env_once(env)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Tool/MSCommon/vc.py", line 695:
    msvc_setup_env(env)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Tool/MSCommon/vc.py", line 814:
    d = msvc_find_valid_batch_script(env,version)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Tool/MSCommon/vc.py", line 762:
    d = script_env(vc_script, args=arg)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Tool/MSCommon/vc.py", line 621:
    stdout = common.get_output(script, args)
  File "c:/msys64/mingw64/lib/python3.8/site-packages/scons/SCons/Tool/MSCommon/common.py", line 238:
    sys.stderr.write(stderr)
```

Alternatively uninstall that scons, and install the msys/scons and try building again:
- `scons platform=windows`
result:
```Bash
$ scons p=windows
scons: Reading SConscript files ...
Package x11 was not found in the pkg-config search path.
Perhaps you should add the directory containing `x11.pc'
to the PKG_CONFIG_PATH environment variable
No package 'x11' found
Configuring for Windows: target=debug, bits=default
Using MinGW
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
[Initial build] build_res_file(["platform/windows/godot_res.windows.tools.64.o"], ["platform/windows/godot_res.rc"])
scons: *** [platform/windows/godot_res.windows.tools.64.o] Error 1
scons: building terminated because of errors.
```

**Minimal reproduction project:**
3.2 branch, hash: ed27b7e6
master 28f0b15c

**Comment by FIF15** [comment]
Just encountered the same problem. I modified the `scons/SCons/Tool/MSCommon/common.py` to print the error correctly and there were multiple lines of "The system cannot find the path specified.". I guess it related to my system environment settings but I didn't want to go deeper on that. Eventually, I just reset the Windows with "remove everything" option, and the problem solved.

**Comment by MichaelBelousov** [comment]
@FIF15 
a) does `--debug=stacktrace` do the same thing for scons?
b) Could you elaborate on what you mean by "reset windows", you mean you performed "Reset this PC" and reinstalled the operating system? I'm not sure that's a [good] option for me.

**Comment by bdbaddog** [comment]
Please try : pip install -i https://test.pypi.org/simple/ SCons==3.9.9a991
That's a fairly recent development build of SCons. 
It's likely your issue has been resolved.

**Comment by MichaelBelousov** [comment]
3.1.1, 3.1.3 fail the same way as 3.1.2
3.9.9a991 fails like so:
```Bash
$ scons
Traceback (most recent call last):
  File "C:/msys64/mingw64/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:/msys64/mingw64/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:/msys64/mingw64/bin/scons.exe/__main__.py", line 4, in <module>
ModuleNotFoundError: No module named 'SCons'
```

**Comment by mwichmann** [comment]
what is `scons.exe/__main__.py`?  this looks like something from msys, not the scons you installed.

**Comment by bdbaddog** [comment]
@MichaelBelousov - that was pip install via msys2's python version > 3.5?

**Comment by MichaelBelousov** [comment]
```Bash
Mike@DESKTOP-KBIIK46 /d/gamedev/debt-civ (master) 20:24:18.348
$ pacman -Qo python
error: No package owns /mingw64/bin/python

Mike@DESKTOP-KBIIK46 /d/gamedev/debt-civ (master) 20:24:25.761
$ python -V
Python 3.8.2

Mike@DESKTOP-KBIIK46 /d/gamedev/debt-civ (master) 20:24:29.091
$ pacman -Qe | grep python
mingw-w64-i686-python-pip 20.0.2-1
mingw-w64-x86_64-python-pip 20.0.2-1

Mike@DESKTOP-KBIIK46 /d/gamedev/debt-civ (master) 20:25:06.783
$ pacman -Qk scons
error: package 'scons' was not found
```
System installed scons (msys2/mingw64/pacman) was uninstalled, I installed using the system pip, which is definitely tied to the one and only python interpreter I have installed on this new system.

**Comment by MichaelBelousov** [comment]
Full example:

```Bash
Mike@DESKTOP /d/gamedev/game (master) 20:27:48.014
$ pacman -R scons
error: target not found: scons

Mike@DESKTOP /d/gamedev/game (master) 20:29:16.329
$ python -V
Python 3.8.2

Mike@DESKTOP /d/gamedev/game (master) 20:29:31.004
$ python -m pip install -i https://test.pypi.org/simple/ Scons==3.9.9a99
Looking in indexes: https://test.pypi.org/simple/
Collecting Scons==3.9.9a99
  Downloading https://test-files.pythonhosted.org/packages/5c/4d/4aa5d7c370d55d3049394b8ada75f9560b5e8f8a6d904d9ebf79fab9754c/SCons-3.9.9a99-py3-none-any.whl (4.0 MB)
     |████████████████████████████████| 4.0 MB 3.3 MB/s
Requirement already satisfied: setuptools in c:/msys64/mingw64/lib/python3.8/site-packages (from Scons==3.9.9a99) (46.1.3)
Installing collected packages: Scons
  Attempting uninstall: Scons
    Found existing installation: scons 3.1.0
    Uninstalling scons-3.1.0:
      Successfully uninstalled scons-3.1.0
Successfully installed Scons-3.9.9a99

Mike@DESKTOP /d/gamedev/game (master) 20:30:20.888
$ scons
Traceback (most recent call last):
  File "C:/msys64/mingw64/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:/msys64/mingw64/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:/msys64/mingw64/bin/scons.exe/__main__.py", line 4, in <module>
ModuleNotFoundError: No module named 'SCons'
```

**Comment by bdbaddog** [comment]
Here's what I get:

```
$ pip install -i https://test.pypi.org/simple/ SCons==3.9.9a991
Looking in indexes: https://test.pypi.org/simple/
Collecting SCons==3.9.9a991
  Downloading https://test-files.pythonhosted.org/packages/56/66/22db0e689b1d3af5b79c7beb56201a2d21debd833498a5eee760b3e22b54/SCons-3.9.9a991-py3-none-any.whl (4.0 MB)
     |################################| 4.0 MB 3.3 MB/s
Requirement already satisfied: setuptools in c:/msys64/mingw64/lib/python3.8/site-packages (from SCons==3.9.9a991) (46.1.3)
Installing collected packages: SCons
Successfully installed SCons-3.9.9a991

Bill@HEXADOG MINGW64 ~
$ which scons
/mingw64/bin/scons

Bill@HEXADOG MINGW64 ~
$ scons --version
SCons by Steven Knight et al.:
        engine: v__VERSION__.__BUILD__, __DATE__, by __DEVELOPER__ on __BUILDSYS__
        engine path: ['C:/msys64/mingw64/lib/python3.8/site-packages/SCons']
__COPYRIGHT__

Bill@HEXADOG MINGW64 ~
$ cd godot/

Bill@HEXADOG MINGW64 ~/godot
$ scons
scons: Reading SConscript files ...

scons: warning: EnsureSConsVersion is ignored for development version
File "C:/msys64/home/Bill/godot/SConstruct", line 3, in <module>
Automatically detected platform: windows
Configuring for Windows: target=debug, bits=default
Found MSVC version 14.2, arch amd64, bits=64
Checking for C header file mntent.h... no
scons: done reading SConscript files.
scons: Building targets ...
[Initial build] scons: *** [core/method_bind.gen.inc] Implicit dependency `core/platform_methods.py' not found, needed by target `core/method_bind.gen.inc'.
scons: building terminated because of errors.
```

**Comment by bdbaddog** [comment]
@MichaelBelousov - any chance you can hop on SCons' discord server or IRC channel?
discord: https://discord.gg/bXVpWAy
IRC: is #scons on freenode

**Comment by MichaelBelousov** [comment]
did a deep dive with @bdbaddog, got past the scons issues by explicitly uninstalling the pip version `pip uninstall scons` then explicitly installing 3.99 via pip again.

Still not past the `Error 1` during build, but it's due to windres landing on:
```
$ windres --include-dir . -i platform/windows/godot_res.rc -o platform/windows/godot_windows.windows.tools.64.o      The system cannot find the path specified.
C:\msys64\mingw64\bin\windres.exe: preprocessing failed.
```

Also note that the file `platform/windows/godot_windows.windows.tools.64.o` doesn't exist, the one that exists in my directory is `platform/windows/godot_res.windows.tools.64.o`

**Comment by MichaelBelousov** [comment]
couple of notes so far:
- `cp platform_methods.py core` satisfied an unfound request for platform_methods.py in a script in core/
- adding the  `--use-temp-file` option to the `windres` command occurring on line 91 of `platform/windows/detect.py` stops windres from erroring out, but there's still the `Error 1` error

**Comment by MichaelBelousov** [comment]
So it looks like that `windres` call is still outputting to stderr according to the POpen.communicate call. When executing:
```
windres --include-dir . --use-temp-file -i platform/windows/godot_res.rc -o platform/windows/godot_res.windows.tools.64.o
```
stdout = None
stderr = b'The system cannot find the path specified.\r\n'

**Comment by MichaelBelousov** [comment]
I assume because of some weird path resolution on msys: removing `shell=True` from the `Popen` call (in addition to the adding of `--use-temp-file` fixed it.

**Comment by MichaelBelousov** [comment]
Also I'm trying to compile with mono support and SCons's `Dir` object fails to detect absolute paths in paths containing back slashes (in 3.99), so I have to manually pass the mono_prefix with forward slashes

**Comment by MichaelBelousov** [comment]
More specifically, Godot gets the mono installation path from the windows registry, and if using msys2, you need to convert that path to a unix path for the msys2 python to work as expected. That might be worth a PR.

**Comment by MichaelBelousov** [comment]
Also looks like the original stack trace about "write() argument must be str, not bytes" was due The build script not handling having visual studio build tools installed, but not the whole visual studio IDE. Supporting the builds tools might also be worthy of a PR.

**Comment by MichaelBelousov** [comment]
Nope, installing the IDE did nothing, I guess I'll fully debug it.

**Comment by FIF15** [comment]
> @FIF15
> a) does `--debug=stacktrace` do the same thing for scons?
> b) Could you elaborate on what you mean by "reset windows", you mean you performed "Reset this PC" and reinstalled the operating system? I'm not sure that's a [good] option for me.

a) I'm not sure about that.
b) Yes, I meant "Reset this PC". I'm sorry this is not a option for you and hope you will find out another way out soon. I tried to solve the problem, but didn't get any clue. So I stopped debugging at some point to move on to the work I planned to do with Godot.

**Comment by avril-gh** [comment]
_i see you trying out different versions of stuff..._

maybe it helps, i have
Win10 x64, MSVC 2019, Scons 2.4.1, Python 2.7.11,
and it works like a charm for both new godot 3.2 and good old 2.1

**Comment by MichaelBelousov** [comment]
Alright, I appear to have found the problem with my environment! In fact, I'll probably go back at some point today and confirm whether any of the changes I made are needed now, and then close the issue.

Firstly, I found that vcvarsall.bat was yielding an error code of 1 and spitting multiple "file not found" lines to stderr while SCons tries to locate vsbuild tools. I opened up VS's developer command prompt. And I noticed that any command prompt I open, at the top it declares "file not found". And I realized that I installed [Clink](https://mridgers.github.io/clink/) for cmd like 4 years ago thinking it would help the times I need to use cmd, but I learned quickly I basically never need to use cmd. Last week I uninstalled it, and I don't believe I uninstalled it in the intelligent way, so I think cmd was broken by missing clink files it expected. I reinstalled clink and it seems to have repaired that, and I can uninstall clink correctly too probably. vsvars is now fine, vs build tools are recognized by scons, and I think that's all I needed, I'll be running builds to see if I run into the windres problem again.
